ENH: properly account for trailing padding in PEP3118

ahaldane opened the pull request:
This changes how numpy treats trailing padding in the PEP3118 buffer interface, to fix #7797.

Previously, dtypes with trailing padding would have their trailing padding truncated when converted to  memoryview using pep3118. Now trailing padding is maintained.

Then, I also needed to modify the pep3118-to-array conversion to stop explicitly adding void fields for the trailing padding. This way code of the form `np.array(memoryview(arr))` will give back the original array even if there is trailing padding.

One thing I still need to think about is what to do for "aligned" types. The pep3118 doc is not clear what "alignment" means. It appears that `ix` and `ixxx` are supposed to mean the same thing: a structure of 8 bytes containing an integer in the first 4 bytes and 4 bytes of padding, but I am not 100% sure that is right.

pv commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5637,24 +5637,27 @@ def test_native_padding_2(self):
-        self._check('ixxxx', [('f0', 'i'), ('', VV(4))])
-        self._check('i7x', [('f0', 'i'), ('', VV(7))])
+        # XXX not sure if this aligment is desired behavior
+        self._check('ix',    dtype_itemsize(8))
```
The alignment in the native mode is supposed to follow that of the C compiler, so the test should check the results are in agreement with `dtype.alignment`

pv commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -292,6 +292,13 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
         }
+        
+        /* Add any trailing padding */
+        while (*offset < descr->elsize) {
```
Maybe should use the prefix-counts, eg  `3x` instead of `xxx`

ahaldane commented:
Thanks @pv, I made some small corrections. 

I need to read and think about the proper behavior for the unstructured types and aligned structured types. 

It seems to me the PEP3118 format supports arrangements that can't really be done in numpy. For example, while numpy *does* support trailing padding in structured types like `T{i:f0:xx}`, there is no natural way in numpy to include trailing padding without a structure, like in `ixx`. Therefore it may be reasonable to raise an error in the latter case, when trying to convert memoryviews to arrays.

Also, I was looking at how the `struct` module treats padding bytes. While it adds internal padding bytes to maintain alignment, it doesn't add trailing padding bytes the way numpy does (in this PR or before):
```python
# b gets 4 bytes here, like in dtype('i,u1,i', align=True)
>>> struct.pack('ibi', 1, 2, 3)  
b'\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00'

# trailing x only adds 1 byte, not 4 as numpy would
>>> struct.pack('ibix', 1, 2, 3)  
b'\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00\x00'

# in numpy (and usually C structs), 4 more trailing pad bytes would be added here
>>> struct.pack('qi', 1, 2) 
b'\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00'
```
The pep3118 docs say the `struct` module is supposed to support the new format strings, but it doesn't appear to yet. So maybe the discrepancy between struct and numpy is struct's problem, not numpy's.

ahaldane commented:
Ah, see

https://bugs.python.org/issue26746
https://bugs.python.org/issue3132

ahaldane commented:
After reading, I think numpy's current interpretation of trailing padding in pep3118 format strings is wrong. The struct module is [pretty clear](https://docs.python.org/2/library/struct.html#byte-order-size-and-alignment) about how format strings treat trailing padding:

>    Padding is only automatically added between successive structure members. No padding is added at the beginning or the end of the encoded struct.
>
>    No padding is added when using non-native size and alignment, e.g. with ‘<’, ‘>’, ‘=’, and ‘!’.
>
>    To align the end of a structure to the alignment requirement of a particular type, end the format with the code for that type with a repeat count of zero. 

So in a format like 'ix', even though the type is implicitly `@` (aligned) we should only add 1 trailing padding byte. If we wanted numpy-style padding bytes the format string should instead be something like 'ix0i'. Unfortunately this means that even though `ix` is in "aligned" mode, the resulting numpy dtype does not count as aligned.

I've updated the code accordingly. There is still more to do.

charris commented:
Might take a look at #7467 while you are working on this.

ahaldane commented:
All right, the story gets more complicated. 

While the struct module does not add C-style trailing bytes to unstructured aligned data, the memoryview function does within `T{}`:
```
>>> arr = np.zeros(1, dtype=np.dtype([('x', '<f8'), ('y', '<b')], align=True))
>>> m1 = memoryview(a)
>>> m1.format, m1.itemsize
('T{d:x:b:y:}', 16)

#  ctypes has the same behavior
>>> class POINT(ctypes.Structure):
...     _fields_ = [("x", ctypes.c_double), ("y", ctypes.c_byte)]
...
>>> m2 = memoryview((POINT)())
>>> m2.format, m2.itemsize
('T{<d:x:<b:y:}', 16)

>>> len(struct.pack('db', 0.0, 0))
9
```

I don't know how to construct a memoryview of format `ix` so I can't test that case. But it appears that trailing padding is added within `T{}` (which does not currently exist in the struct module) while struct does not add it outside of `T{}`.

aldanor commented:
@ahaldane Just wondering, are there any plans in regards to moving this forward?

ahaldane commented:
Yeah I sort of forgot about this because it was so messy.

As I recall I think the current state of this PR is a pretty good approximation of how things "should work", given the current somewhat incomplete state of pep3118 and memoryviews in python. 

I'll revisit it sometime soon, comments on it currently are welcome.

eric-wieser commented:
Some overlap here with #9054

ahaldane commented:
Rebased. 

As a reminder, this PR fixes the PEP3118 trailing padding, which was not accounted for at all before when converting an `array->memoryview`. Importantly, I now interpret `ix` format as 5 bytes, and `ix0i` as 8 (on x64), as the struct modele documents should be the case. This means I had to tweak the `memoryview->array` code since it did something else before. This should have no effect on any real code since such formats are never created in practice.

ahaldane commented:
Here are two more things we might want to think more about:

1. Currently, although the format string numpy produces from an array works properly, is perhaps a bit ugly. Consider:
```python
>>> memoryview(np.zeros(3, dtype=np.dtype('u1,i4,u1', align=True))).format
'T{B:f0:3xi:f1:B:f2:3x}'
```
Maybe it should be `T{B:f0:0ii:f1:B:f2:0i}`.

2. While fixing this up I realized there are dtypes which we will never be able to represent with the PEP3118 sytnax: Those with overlapping fields:
```python
>>> dt = dtype({'names': ['a', 'b'], 'formats': ['i4','i4'], 'offsets': [0,1]})
>>> memoryview(np.zeros(3, dtype=d))
RuntimeError: This should never happen: Invalid offset in buffer format string g
eneration. Please report a bug to the Numpy developers.
```

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -248,14 +260,17 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
-                                "buffer format string generation. Please "
-                                "report a bug to the Numpy developers.");
+                                "The buffer interface does not support "
+                                "overlapping fields");
```
Doesn't this path also trigger for fields that are not in sequence?

ahaldane replied:
For future reference, one more datapoint: The HDF5 datatype specification also stores "compound datatypes" as an ordered list of `(name, format, offset)` tuples. So that is now 3 cases where structured datatypes are stored that way: numpy dtypes, MPI serialization, and HDF5 files. 

I want to record these examples here in case in the future we want to argue to the CPython devs that we should add a new PEP3118 specification style for structured types, that is equivalent to a list of tuples.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5918,6 +5921,10 @@ def aligned(n):
 
+        expected_dtype = {'names': ['f0'], 'formats': ['i'],
+                          'itemsize': np.dtype('i,V1', align=True).itemsize}
+        self._check('(3)T{ix}', (expected_dtype, (3,)))
```
Isn't this an exact duplicate of the above test? What am I missing?

ahaldane replied:
oh sorry, I fudged the rebase here

eric-wieser replied:
Although perhaps using `np.dtype(..., align=True)` is an improvement over `aligned(size + 1)`

eric-wieser commented:
What is the itemsize of `T{ix}` with this patch?

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5856,28 +5856,31 @@ def test_native_padding_2(self):
-        self._check('ixx',   dict(itemsize=aligned(size + 2), **base))
-        self._check('ixxx',  dict(itemsize=aligned(size + 3), **base))
-        self._check('ixxxx', dict(itemsize=aligned(size + 4), **base))
-        self._check('i7x',   dict(itemsize=aligned(size + 7), **base))
```
I think these tests might remain valid with `T{...}`? In particular, I would expect trailing padding to be kept in a struct context, to match the behaviour of `sizeof(T)` in C, and what happens when structs are repeated

eric-wieser replied:
Whereas the `ctypes` modules has set the precedent of ignoring all the remarks that the struct docs make about alignment...

ahaldane commented:
I expect `T{ix}` is 5 bytes. However I currently do not know of any way (in python) to create a memoryview with that format.

It *is* possible to create a `T{=ix}` format, though (in this PR):
```python
>>> a = np.zeros(3, dtype=np.dtype({'names': 'a', 'formats': 'i4', 'itemsize':5}))
>>> memoryview(a).format
'T{=i:a:x}'
>>> np.array(memoryview(a))
array([(0,), (0,), (0,)],
      dtype={'names':['a'], 'formats':['<i4'], 'offsets':[0], 'itemsize':5})
```

eric-wieser commented:
For anyone else reading this, a relevant python bug/discussion is here: https://bugs.python.org/issue26746

ahaldane commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -248,14 +260,17 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
-                                "buffer format string generation. Please "
-                                "report a bug to the Numpy developers.");
+                                "The buffer interface does not support "
+                                "overlapping fields");
```
yes, I should make a better message

eric-wieser replied:
What are we going to do about #6053 here? Seems like it's going to make casting to memoryview a lot harder...

ahaldane replied:
Oh I see, you're right. The PEP3118 format cannot account for either overlapping fields *or* out-of-order fields (when order matters). That's something to think about...

eric-wieser replied:
I'm a little confused by PEP3118, as it:

* Seems to be designed to standardize some numpy behaviour for interoperability
* Causes discussions in the python bugtracker that refer to numpy as the canonical implementation
* Is not actually sufficient to represent either all C types or all numpy types

ahaldane replied:
Just a thought (weeks later),

A more general way of describing a structured type is as a list of `(name, type, offset)` tuples. This is how, for example, MPI serializes structs (see [here](http://beige.ucs.indiana.edu/I590/node100.html)). It is also how numpy already stores structured types internally.

Since a purpose of PEP3118 is serialization, maybe it would have made sense to specify structured types this way. Eg we could have had something like `'T24{i,0,f0:f,8,f1}` to represent `dtype({'formats': ['i', 'f'], 'offsets': [0, 8], 'names': ['f0', 'f1'], 'itemsize': 24})`.

Since numpy is the main user of the PEP3118 interface, and because the PEP seems to be only partly implemented in cpython currently, there is probably still opportunity for us to work out the details or add to the specification, with some effort.

ahaldane commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5856,28 +5856,31 @@ def test_native_padding_2(self):
-        self._check('ixx',   dict(itemsize=aligned(size + 2), **base))
-        self._check('ixxx',  dict(itemsize=aligned(size + 3), **base))
-        self._check('ixxxx', dict(itemsize=aligned(size + 4), **base))
-        self._check('i7x',   dict(itemsize=aligned(size + 7), **base))
```
The PEP3118  spec is unclear about this.

One could argue the `struct` module has set no precedent in judging how alignment and padding work here since it doesn't implement the `T{}` format. We might then feel free to set the precedent here in this PR, deciding that aligned formats add trailing padding only inside `T{}`.

charris commented:
@seberg @eric-wieser  Is this still of interest?

charris commented:
Needs rebase. I'm inclined to close this unless it is still relevant. @seberg @eric-wieser thoughts?

seberg commented:
It is still relevant, there was no change in behaviour with respect to the issue.

I am not sure whether the trailing bytes themselves should or should not be relevant. The main interest would be with respect to getting alignment right in the buffer protocol (this PR does that as far as I can tell).  But the mismatch between buffer protocol itemsize and the `length` that is set, does currently prevent roundtripping `array -> buffer -> array` in certain cases, so I guess there is something wrong.

(I am not sure how easy it is to rebase this PR, it might be annoying, but I expect all the pieces here are still relevant just as much)

ahaldane committed:
ENH: output trailing padding in PEP3118 format strings

Fixes #7797

ahaldane committed:
ENH: allow 0-sized elements in PEP3118 format strings to align

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -276,15 +289,20 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
             /* Insert padding manually */
             if (*offset > new_offset) {
                 PyErr_SetString(
-                    PyExc_ValueError,
-                    "dtypes with overlapping or out-of-order fields are not "
-                    "representable as buffers. Consider reordering the fields."
-                );
+                    PyExc_ValueError, "The buffer interface does not support "
+                                      "overlapping fields or out-of-order "
+                                      "fields");
```
Is this change of message deliberate, or just a consequence of the merge?

ahaldane replied:
deliberate, based on your review above.

I just did a little bit more than a simple rebase. I also added support for 0-sized unnamed padding, see the release note.

ahaldane replied:
oh I see what you mean, the previous message seems good enough.

ahaldane commented:
* rebased
 * tidied up based on review comments from last time
 * added release note
 * NEW: disallow automatic assumption of trailing padding if the format is *not* inside `T{}`, as that is what the struct doc says to do.
 * NEW: added support for 0-sized unnamed fields (eg to add trailing padding), as struct supports.

See: [struct doc](https://docs.python.org/3/library/struct.html) in particular "Note" #3 and the last example.

aldanor commented:
@ahaldane It's been quite a while, just wondering what's the status on this?

ahaldane commented:
As far as I remember, this is good-to-go, after review. 

I can rebase and re-review it myself in the next few days. As I recall the challenge was in figuring out a consistent way to interpret the various python/numpy docs.

InessaPawson commented:
@ahaldane Please let us know if you need help with moving this PR forward.

mattip commented:
We discussed this at a recent triage meeting and decided to close it due to a lack of interest in moving it forward. If someone wants to pick it up, we would review the work. Thanks @ahaldane for looking at this.
